### ISSUE TITLE
Add hard triplet miner layer for efficient triplet network implementations

### ISSUE THREAD

**Issue by maxsch3** [tensorflow/addons#837]
I'm proposing to add a new layer for building efficient triplet networks that use approach described in this paper: https://arxiv.org/abs/1703.07737. 

In summary, the paper points out that naive triplet generation when triplets of samples are selected randomly before feeding them into a network is not efficient. They suggest that triplets can be 'mined' inside a network which has two benefits: 

- computational: in each batch, a datapoint is passed through the encoder network only once and then it can be reused multiple times in different triplets (in combinations with other datapoints)
- better convergence: by having encoded representation at a time of forming triplets, one can pick so called 'hard' triplets - the worst combinations in terms of triplet loss. This ensures that triplet loss function always feeds good gradients into optimiser

I have already used this approach very successfully in couple of my projects as custom layer and Now I would like to share it with the community.  

Layer inputs:

- encoded representation of a datapoint (embedding)
- datapoint label

Layer outputs:

- a tensor of dimensions (None, 2). First column will hold cosine similarities of hard positive pairs
mined by the layer. The second column will hold the same for hard negative pairs mined.
 
This tensor can then be plugged into a triplet loss function which will penalise low cosine similarities in positive pairs and high similarity in negative ones. From my experience custom loss function can vary slightly from project to project, so it is good to leave freedom in making this function to developers. 

The suggested starting point for a triplet loss function is : L = logcosh(x_pos - 1) + exp(5*(x_neg-1)) see plots on wolfram alpha https://www.wolframalpha.com/input/?i=+log%28cosh%28x-1%29%29%2C+exp%285*%28x-1%29%29+from+-1+to+1

**Relevant information**
- Are you willing to contribute it (yes/no): yes
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where): yes
- Is there already an implementation in another framework? (if so, where): not sure
- Was it part of tf.contrib? (if so, where): no

**Which API type would this fall under (layer, metric, optimizer, etc.)**
Layer

**Who will benefit with this feature?**
Anyone who wants to build a triplet network

**Any other info.**

**Comment by seanpmorgan** [comment]
Thanks for the suggestion @maxsch3 . I think this fits nicely. Please submit a PR. If you could include a tutorial since you know models it can be used in that be ideal.

**Comment by maxsch3** [comment]
Ah. Sorry. I had a second look and I found very similar functionality is implemented already as a loss in this repo. 

Closing as not needed
